frontend/src/utils/queryGraphUtils.js

```diff
@@ -112,10 +113,51 @@ const convert = {
   },
 };
 
+/**
+ * Remove any empty node categories or ids or edge predicates
+ * @param {obj} q_graph query graph
+ * @returns query graph
+ */
+function prune(q_graph) {
+  const clonedQueryGraph = _.cloneDeep(q_graph);
+  Object.keys(clonedQueryGraph.nodes).forEach((n) => {
+    pruneEmptyArrays(clonedQueryGraph.nodes[n], 'category');
+    pruneEmptyArrays(clonedQueryGraph.nodes[n], 'id');
+    if (clonedQueryGraph.nodes[n].id && clonedQueryGraph.nodes[n].category) {
+      delete clonedQueryGraph.nodes[n].category;
+    }
+  });
+  Object.keys(clonedQueryGraph.edges).forEach((e) => {
+    pruneEmptyArrays(clonedQueryGraph.edges[e], 'predicate');
+  });
+  return clonedQueryGraph;
+}
+
+function ingest(q_graph) {
+  const clonedQueryGraph = _.cloneDeep(q_graph);
+  Object.keys(clonedQueryGraph.nodes).forEach((n) => {
+    const node = clonedQueryGraph.nodes[n];
+    standardizeIDs(node);
```

Comment:
This to me looks like a clear case for a more object-oriented approach.

Reply:
This function is only used when importing a query graph from the json editor to change whatever the user put in to a more standard format. Once it's in that standard format, we don't have to do this whenever changes are made in the UI because we can control those changes.